Post Checkout: Don't open internal links in a new window

drewblaisdell opened the pull request:
Fixes #2810.

<img width="975" alt="screen shot 2016-01-26 at 2 50 41 pm" src="https://cloud.githubusercontent.com/assets/1130674/12600624/169adf96-c450-11e5-822d-3f1515cad80c.png">

This PR adds a commit to change the API of `PurchaseDetail` to accept props `href` and `target` instead of `onButtonClick`. This is an improvement over the previous method (using a single `onClick` prop) because:

- A new function isn't created during every render().
- It allows us to open links in the same window by omitting `target`.

All internal links should open in the same window now.

**Testing**
We should assert that the links highlighted above work on the thank you pages when checking out with the following products:

```
Business Plan
Chargeback
Domain Mapping
Domain Registration
Google Apps
Jetpack Business Plan
Jetpack Premium Plan
Premium Plan
Site Redirect 
```

A few of these might be difficult to test (I'm unsure about the chargeback and the Jetpack plans), but you can test the thank you for page for them by adding `componentClass = [thank you component]` on [this line](https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/upgrades/checkout/thank-you.jsx#L242) where `[thank you component]` is one of the component names above, without spaces, followed by 'Details' (e.g. `JetpackBusinessPlanDetails`), then checking out with any product.

This builds on commits from #2713 which need to be removed when that is merged, so I'm labeling this as 'In Progress' until that is merged.

- [x] Product review
- [x] Code review

fabianapsimoes commented:
I found some issues here that are probably not related to this PR. Feel free to open them in separate issues.

- In the Thank You page for the Business plan, Browse Themes points to `/themes/:site`. This route doesn't seem to work. It should point to `/design/:site`.

- I noticed that plan information isn't re-fetched after checkout. Because of that, for example, if I reach the Thank You page for Business and click the eCommerce option, I'll be prompted to upgrade instead of see the list of plugins. I need to refresh in order for Calypso to "see" I have a plan.

- Not really an issue, but I do not see the Thank You page for Themes. Should I? This is what I see:
![screenshot-calypso localhost 3000 2016-01-28 12-40-13](https://cloud.githubusercontent.com/assets/1248436/12643419/54a068c0-c5bc-11e5-89ba-81aaf5cd272f.png)

Also, pinging @v18 and @jeherve for feedback on Jetpack Plans.

> A few of these might be difficult to test (I'm unsure about the chargeback and the Jetpack plans), but you can test the thank you for page for them by adding componentClass = [thank you component] on this line where [thank you component] is one of the component names above, without spaces, followed by 'Details' (e.g. JetpackBusinessPlanDetails).

Which route am I supposed to visit? I did that and then visited `/checkout/thank-you` and `/checkout/thank-you/:site`, but I couldn't see the Thank You page.

stephanethomas commented on client/my-sites/upgrades/checkout/thank-you.jsx:
```diff
@@ -346,14 +349,16 @@ JetpackPremiumPlanDetails = React.createClass( {
 					description={ this.translate( 'Say goodbye to comment spam' ) }
 					buttonText={ this.translate( 'Start using Akismet' ) }
-					onButtonClick={ goToExternalPage( 'https://en.support.wordpress.com/setting-up-premium-services/' ) } />
+					href="https://en.support.wordpress.com/setting-up-premium-services/"
```
I know this is not the purpose of this pull request but we could use this opportunity to remove the `en` locale prefix from these support links.

stephanethomas commented:
> Not really an issue, but I do not see the Thank You page for Themes. Should I?

Maybe @ockham or @mcsf might know that.

drewblaisdell commented:
> - In the Thank You page for the Business plan, Browse Themes points to `/themes/:site`. This route doesn't seem to work. It should point to `/design/:site`.

This has been updated.

> I noticed that plan information isn't re-fetched after checkout. Because of that, for example, if I reach the Thank You page for Business and click the eCommerce option, I'll be prompted to upgrade instead of see the list of plugins. I need to refresh in order for Calypso to "see" I have a plan.

We had an issue for this related to free trials, but I updated it for all plans, as the solution shouldn't change (#2818). That issue should probably be a blocker for this, as it is exaggerated when the link opens in the same window.

> Which route am I supposed to visit? I did that and then visited `/checkout/thank-you` and `/checkout/thank-you/:site`, but I couldn't see the Thank You page.

Once you edit that component, you go through the checkout with any product and are taken to the 'Thank You' page for the product you hardcoded into `ThankYou`. I updated the testing instructions to include this info.

drewblaisdell commented:
This is no longer blocked by #2908.

stephanethomas commented:
Code looks good. I took advantage of this pull request to remove unused code and fix an issue with localized links.

It's worth noting that there might be a usability issue with this new approach since users can't go back to the `Thank You` page once they clicked on any of its links - they are redirected on the `Plans` page instead.

drewblaisdell commented:
Thanks @stephanethomas, those changes look good to me.

scruffian commented:
As @stephanethomas mentioned, this breaks the behaviour of the back button. To fix it we'll probably need to address this:

https://github.com/Automattic/wp-calypso/issues/2235

I would consider #2235 a blocker for this PR - many users won't open those links in a new tab, and would expect to be able to get back to them.

drewblaisdell commented:
> I would consider #2235 a blocker for this PR - many users won't open those links in a new tab, and would expect to be able to get back to them.

Sounds good. One note: going back to the thank you page is broken in master (worse than in this PR, because an error is thrown) as well.

fabianapsimoes commented:
> Sounds good. One note: going back to the thank you page is broken in master (worse than in this PR, because an error is thrown, instead of redirecting the user to /plans) as well.

Considering this, I'm not sure we should block this PR on #2235. I'd suggest merging it, and addressing the Back button separately later.

drewblaisdell commented:
@fabianapsimoes Sorry, my comment might have made this unclear (or you know this and don't think it's a regression): the back button is broken in master/this branch, but @scruffian / @stephanethomas have pointed out that it is more noticeable here because the links now don't open in a new window.

fabianapsimoes commented:
@drewblaisdell thanks for clarifying! That makes sense then.

scruffian commented:
While testing I found this:
<img width="392" alt="screen shot 2016-02-19 at 23 32 04" src="https://cloud.githubusercontent.com/assets/275961/13192310/8551f756-d762-11e5-8569-01990a9f4aee.png">

drewblaisdell commented:
Thanks for catching that, @scruffian. I added a commit to fix the loading state of the support messages.

scruffian commented:
LGTM :+1:
